AWAE: Mod_AmbWind=3 add error if HR grid not centered on turbine in Y dimension

andrew-platt opened the pull request:
This is ready to merge.

**Feature or improvement description**
When the boundaries for _InflowWind_ interpolation were tightened to end at the edge of the grid with version 3.5.0, this had the unintended consequence of not allowing a small shift of the turbine from the exact center of the high-resolution grid.  Previously extrapolation of the wind data outside the grid in the +/-Y direction of up to one step was permitted which allowed the turbine to be slightly offset for `Mod_AmbWind=3` cases.  However, this is in direct conflict with our modeling guidance for `Mod_AmbWind=3` cases; our guidance indicates the high resolution grid must exactly match the _InflowWind_ grid.

The following diagram illustrates what the issue is:
![FF_highResGrid_new_offset](https://github.com/OpenFAST/openfast/assets/2745453/91d3f897-0581-4e0a-818e-386687cb5b0a)
Figure: FAST.Farm high res grid and OpenFAST wind grid (`Mod_AmbWind=3`) in v3.5.x. This is the case where the grid sizing and spacing for the _InflowWind_ grid and high-res grid are identical, but the turbine is slightly offset.  The _InflowWind_ grid is always centered about the turbine and therefore does not line up with the high-resolution grid. This effectively moves points in the high-resolution grid outside the available interpolation boundary for the _InflowWind_ wind field.

So to increase usability, we will leave the stricter boundaries in the _InflowWind_ grid interpolation.  To make it possible for an end user to figure out what failed, we add an error message indicating what exactly failed and how to fix it.  Example error message:
![Screenshot 2024-01-08 at 7 47 25 PM](https://github.com/OpenFAST/openfast/assets/2745453/35425699-1a7a-44e1-9f85-766048a74eb7)




**Related issue, if one exists**
#1951 
https://forums.nrel.gov/t/gf-wind-array-boundaries-violated-grid-too-small-in-y-direction/6149/5

**Impacted areas of the software**
FAST.Farm usability is somewhat enhanced for `Mod_AmbWind=3` cases.

**Additional supporting information**
There had been multiple reports of issues with `Mod_AmbWind = 3` when a turbine was not aligned exactly with the center of high-resolution grid in the Y direction. 

**Test results, if applicable**
N/A

jjonkman reviewed (commented):
Overall, this logic makes sense to me.  But I do have a few concerns:
1) With this change, we wont allow the case where the high-res domain is a subset of the TurbSim domain.  I'm not sure the user would ever want to do this, but this functionality was possible before this change.
2) This change forces wind to always propagate along X.  The FAST.Farm documentation also suggests that you could propagate wind along Y (by setting PropagationDir in InflowWind to +/-90deg), although I'm not sure anyone ever does this in practice.
3) Is there a reason why the Z boundary is not checked in addition to Y?

andrew-platt commented:
Good thoughts.

1. I was somehow thinking a smaller grid wasn't possible.  I'll  can modify the checks to allow smaller grids.
2. I wasn't aware of this feature.  This should be relatively simple to check.
3. Completely forgot to add this. I'll make this change also.

andrew-platt commented:
Updated with the above suggestions.  Also added a check to make sure the `PropagationDir` from _InflowWind_ is 0, +/-90, or 180.

New error messages (shifted high-res in both Y and Z for test):
![Screenshot 2024-01-10 at 5 08 15 PM](https://github.com/OpenFAST/openfast/assets/2745453/f4c5b55d-7a49-428d-8436-c49e95ca4418)

jjonkman commented on modules/awae/src/AWAE.f90:
```diff
@@ -1258,6 +1265,88 @@ subroutine AWAE_Init( InitInp, u, p, x, xd, z, OtherState, y, m, Interval, InitO
+      ff  => p%IfW(nt)%FlowField 
+      wfi => IfW_InitOut%WindFileInfo
+
+      tmpMsg = NewLine//NewLine//'Turbine '//trim(Num2LStr(nt))//' -- Mod_AmbWind=3 requires the InflowWind high-res grid '// &
```
I believe you mean to say that the FAST.Farm high-res grid must be entirely contained within the InflowWind flow field.

andrew-platt replied:
Yes.  That is much better wording.  Changed.

jjonkman reviewed (commented):
Looking better, but here are a few suggested corrections...

abhineet-gupta commented:
The error message gets triggered for my case which is helpful. 
One possible improvement we could make is to add the turbine number in the messages too.
Perhaps something like:
```
 Try setting:
     Y0_high = 90
     dY_High = 5
 for turine # 1
```

andrew-platt commented on modules/awae/src/AWAE.f90:
```diff
@@ -1258,6 +1265,88 @@ subroutine AWAE_Init( InitInp, u, p, x, xd, z, OtherState, y, m, Interval, InitO
+      if (wfi%YRange_Limited) then
+         ! flow field limits (with grid tolerance)
+         ff_lim(1) = p%WT_Position(2,nt) + wfi%YRange(1) - GridTol
+         ff_lim(2) = p%WT_Position(2,nt) + wfi%YRange(2) + GridTol
```
Good catch.  
Corrected.

andrew-platt commented on modules/awae/src/AWAE.f90:
```diff
@@ -1258,6 +1265,88 @@ subroutine AWAE_Init( InitInp, u, p, x, xd, z, OtherState, y, m, Interval, InitO
+         ff_lim(2) = p%WT_Position(2,nt) + wfi%YRange(2) + GridTol
+
+         ! wind X aligned with high-res X
+         if (.not. ff%RotateWindBox) then
```
Good catch.  I corrected this as well.

andrew-platt commented:
@abhineet-gupta, good suggestion.  The turbine number is listed just above the suggested grid change, but may not be obvious.  I made this change.
